src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs

```diff
@@ -4808,23 +4808,33 @@ private void VisitDeconstructionArguments(ArrayBuilder<DeconstructionVariable> v
             if (!conversion.DeconstructionInfo.IsDefault)
             {
                 VisitRvalue(right);
+                var rightResult = ResultType;
+                var rightResultWithAnnotations = rightResult.ToTypeWithAnnotations();
 
                 var invocation = conversion.DeconstructionInfo.Invocation as BoundCall;
                 var deconstructMethod = invocation?.Method;
 
                 if ((object)deconstructMethod != null)
                 {
-                    if (!invocation.InvokedAsExtensionMethod)
-                    {
-                        _ = CheckPossibleNullReceiver(right);
-                    }
-                    else
+                    _ = CheckPossibleNullReceiver(right);
+
+                    if (invocation.InvokedAsExtensionMethod && invocation.Method.IsGenericMethod)
                     {
-                        // https://github.com/dotnet/roslyn/issues/33006: Check nullability of `this` argument.
-                    }
+                        // re-infer the deconstruct parameters based on the 'this' parameter 
+                        ArrayBuilder<BoundExpression> placeholderArgs = ArrayBuilder<BoundExpression>.GetInstance(n);
+                        placeholderArgs.Add(CreatePlaceholderIfNecessary(right, rightResultWithAnnotations));
+                        for (int i = 0; i < n; i++)
+                        {
+                            placeholderArgs.Add(new BoundExpressionWithNullability(variables[i].Expression.Syntax, variables[i].Expression, NullableAnnotation.Oblivious, conversion.DeconstructionInfo.OutputPlaceholders[i].Type));
+                        }
+                        deconstructMethod = InferMethodTypeArguments(invocation, deconstructMethod, placeholderArgs.ToImmutableAndFree());
 
-                    // https://github.com/dotnet/roslyn/issues/33006: Update `Deconstruct` method
```

Comment:
>Update `Deconstruct` method [](start = 70, length = 27)

Are we updating the `Deconstruct` method when it is an instance method?

For instance, in the following, we should not report a warning for `x.ToString()`.

```
class Pair<T, U>
{
    public void Deconstruct(out T t, out U u) => throw null!;
}

class Program
{
    static Pair<T, U> CreatePair<T, U>(T t, U u) => new Pair<T, U>();

    static void F(string? x, object? y)
    {
        if (x == null) return;
        var p = CreatePair(x, y);
        (x, y) = p;
        x.ToString(); // ok
        y.ToString(); // warning
    }
}
```

Reply:
Fixed and added Deconstruction_30 to cover this case

---
In reply to: [279097281](https://github.com/dotnet/roslyn/pull/35016#discussion_r279097281) [](ancestors = 279097281)